### DISCUSSION TITLE
feat: include an essentiality summary for each release

### DISCUSSION THREAD

**Discussion by haowang-bioinfo** [SysBioChalmers/Human-GEM#390] [Ideas]
Inspired by this [PR](https://github.com/SysBioChalmers/yeast-GEM/pull/306) from yeast-GEM, it could be nice to include and update an essentiality summary for each `Human-GEM` release.

**Comment by JonathanRob** [comment]
I can see the benefit of such a test, but there are key differences between yeast and human that make this difficult to implement in practice.

The three major challenges that come to mind are:
1. Unicellular vs. multicellular organism
2. Ability to grow on minimal medium
3. Feasibility of experimental procedures

**Unicellular vs. multicellular organism**: Yeast are unicellular, so an essential gene can be defined relatively easily as a gene that is necessary for a single cell to remain viable. Not only do humans consist of many cells, but these are different cell _types_ as well. So then how to define an essential gene? Essential for any cell type? Some types? All types? Essential for an entire tissue, or organism, or just the single cells? There is no right or wrong answer, so it becomes very subjective and dependent on context.

**Ability to grow on minimal medium**: Yeast can grow using defined, minimal media containing a single carbon source (with some other salts, vitamins, etc.). This allows for more controlled tests of gene essentiality, where one can define genes as essential when they are required for yeast viability when grown on that media. Mammalian cells are comparatively difficult to grow, and require complex growth media with amino acids, essential fatty acids, etc., which adds complexity to the problem. Both yeast and human cells will share the complication that gene essentiality is relative to the media they're grown in. So the set of essential genes for yeast grown in minimal media with glucose will be different than when yeast is grown in media supplemented with amino acids, for example.

**Feasibility of experimental procedures**: We cannot perform true organism-level gene essentiality tests on humans (for obvious reasons), which is otherwise relatively simple for yeast. We therefore need to rely on cell line assays, and this requires the assumption that essential genes identified for a cell line will translate to in vivo.

So at least these points would need to be considered when designing a "standard test" for gene essentiality. And keep in mind that using a specific dataset/condition/cell line can lead to "overfitting" the model on that dataset, so the more general, the better.

**Comment by haowang-bioinfo** [comment]
@JonathanRob very good and thorough considerations!

> 1. Unicellular vs. multicellular organism
> 2. Ability to grow on minimal medium
> 3. Feasibility of experimental procedures

As you correctly pointed out that it's impossible to `perform true organism-level gene essentiality tests on humans`, a feasible way might start to collect essential genes for specific tissues and cell lines with high-quality omics data, as well as associated medium conditions. What do you think?

**Reply by mihai-sysbio** [reply]
To me, what is more interesting it to have the technical infrastructure to set this up, and add gene essentiality tests gradually.

**Reply by haowang-bioinfo** [reply]
agree

**Reply by feiranl** [reply]
Agree, maybe also to ensure the metabolic tasks can fill with each release and also the energy and redox production are realistic? For now, it can be run manually before each release. If we find a good way to automate that, then we can incorporate that into the check function

**Reply by mihai-sysbio** [reply]
We can already run actions with [Python code](https://github.com/SysBioChalmers/Human-GEM/blob/7df306201cc7fbfe329ce801d4a1db12ef3de15c/.github/workflows/yaml-validation.yml#L32) and [Matlab code](https://github.com/SysBioChalmers/Human-GEM/blob/7df306201cc7fbfe329ce801d4a1db12ef3de15c/.github/workflows/yaml-conversion.yml#L17). As long as tests are nicely scripted, they can be easily turned into actions.

**Reply by haowang-bioinfo** [reply]
> maybe also to ensure the metabolic tasks can fill with each release and also the energy and redox production are realistic

this might be an urgent thing because the model may easily get broken due to high amount of changes or even accidental mistakes, it would be better to ensure these in each PR

**Reply by mihai-sysbio** [reply]
How should the output be shared? As a comment on the PR?

**Reply by haowang-bioinfo** [reply]
output can be in evaluated by such a [function](https://github.com/SysBioChalmers/Human-GEM/commit/39dc3c54ba1f2ff351604e3e8f82560cecf61779)

**Comment by haowang-bioinfo** [comment]
A practical implementation probably is to begin with the [Hart2015](https://www.cell.com/cell/fulltext/S0092-8674(15)01495-6) datasets which are:

1. homogenous cell lines
2. able to grow on defined medium (McCoy's 5A, DMEM, or Neurocult NS-A Basal media)
3. with fitness genes detected by high-resolution CRISPR knockout and screening procedures

**Reply by haowang-bioinfo** [reply]
Here's small scale essentiality evaluation of combined Hart2015 datasets (`all`), by reusing the code of Human1 paper.

| version | TP | TN   | FP  | FN | accuracy          | sensitivity       | specificity       | F1                | MCC               |
|---------|----|------|-----|----|-------------------|-------------------|-------------------|-------------------|-------------------|
| v1.12   | 40 | 2333 | 175 | 77 | 0.904000000000000 | 0.341880341880342 | 0.930223285486443 | 0.240963855421687 | 0.204768560393159 |
| v1.13   | 40 | 2334 | 174 | 77 | 0.904380952380952 | 0.341880341880342 | 0.930622009569378 | 0.241691842900302 | 0.205504558241293 |
| v1.14   | 40 | 2334 | 175 | 77 | 0.904036557501904 | 0.341880341880342 | 0.930251096054205 | 0.240963855421687 | 0.204787829413435 |

It shows that the performance in essentiality stays relatively stable during recent GPR curation.

**Reply by mihai-sysbio** [reply]
> Here's small scale essentiality evaluation of combined Hart2015 datasets (`all`), by reusing the code of Human1 paper.

Interesting, the reported MCC score with Hart2015 is probably similar to the lowest with DepMap (Fig 4C in the manuscript).

And out of curiosity, how many genes in Hart2015 are in Human-GEM? I'm thinking of contrasting the number of genes we are "testing" vs the rest.

**Reply by haowang-bioinfo** [reply]
> how many genes in Hart2015 are in Human-GEM?

The Hart2015 datasets are from genome-scale CRISPR analysis and nearly cover all human genes (that are presented in Table S2 with gene symbols), which therefore basically include entire Human-GEM genes.

**Reply by haowang-bioinfo** [reply]
here's the essentiality evaluation using the latest model in `develop` branch, it shows nice improvements particularly in decreased false positive values (175 -> 167) and correspondingly imporved true negatives (2334 -> 2343), validating the solid progress from recent community curation since v1.14.

| version | TP | TN   | FP  | FN | accuracy          | sensitivity       | specificity       | F1                | MCC               |
|---------|----|------|-----|----|-------------------|-------------------|-------------------|-------------------|-------------------|
| 2df114603c95834e314fe8be0b91fbd8cf92f061 | 40 | 2343 | 167 | 77 | 0.907118385991625 | 0.341880341880342 | 0.933466135458167 | 0.246913580246914 | 0.210822258032951 |

**Reply by mihai-sysbio** [reply]
It would be nice to map the changes in TN, FP with the GPR curation. Specifically, I'm curious how come the TN goes +1 and then -1.

It's very nice to see that the FP is going down.

**Reply by haowang-bioinfo** [reply]
> It would be nice to map the changes in TN, FP with the GPR curation

good idea

**Reply by mihai-sysbio** [reply]
> > how many genes in Hart2015 are in Human-GEM?
> 
> The Hart2015 datasets are from genome-scale CRISPR analysis and nearly cover all human genes (that are presented in Table S2 with gene symbols), which therefore basically include entire Human-GEM genes.

Indeed, this dataset covers all genes in Human-GEM. From a coverage perspective, it's perfect!

**Reply by mihai-sysbio** [reply]
There are however about 4 or so genes without symbols, so of course those cannot be mapped automatically. I haven't looked into these exceptions.

**Reply by haowang-bioinfo** [reply]
In the latest `genes.tsv` each gene has a unique symbol assigned, please review them in #549 and correct me if anything is wrong

**Reply by mihai-sysbio** [reply]
I didn't look any deeper, but here are the genes with no symbols

ENSG00000100101
ENSG00000255730
ENSG00000259916
ENSG00000272916
ENSG00000284844
ENSG00000285043
ENSG00000285269

**Reply by haowang-bioinfo** [reply]
> here are the genes with no symbols

are these genes in the latest genes.tsv or the model?

**Reply by mihai-sysbio** [reply]
These are the IDs that have no symbols in the TSV file.

**Reply by JonathanRob** [reply]
Thanks for checking @mihai-sysbio. Just a reminder @haowang-bioinfo that Ensembl IDs and gene symbols are never guaranteed to be 1:1. Sometimes they are 1:many, many:1, 1:0, 0:1, etc.

**Reply by haowang-bioinfo** [reply]
> Just a reminder @haowang-bioinfo that Ensembl IDs and gene symbols are never guaranteed to be 1:1. Sometimes they are 1:many, many:1, 1:0, 0:1, etc.

@JonathanRob sometimes yes. Currently the symbols in genes.tsv are extracted from Ensembl by the built-in python [script](https://github.com/SysBioChalmers/Human-GEM/blob/main/code/GPRs/fetch_ensembl_gene_annotations.py). So far they are congruent with Uniprot mapping. In the latest model and genes.tsv, all symbols are uniquely assigned as 1:1 pair

**Comment by feiranl** [comment]
Also check wether there are duplicate rxns for each release to avoid the error identified in #561 #535, and it would also be nice if we can check wether the added rxns are mass balanced and charge balanced. I can work on that.

**Reply by mihai-sysbio** [reply]
Since there is already access to Memote in the Docker image, I would recommend reusing as many of the Memote tests as possible (as previously mentioned [here](https://github.com/SysBioChalmers/Human-GEM/pull/556#issuecomment-1541364128)), for example [test_reaction_charge_balance](https://memote.readthedocs.io/en/latest/autoapi/test_consistency/index.html?highlight=charge#test_consistency.test_reaction_charge_balance) and [test_reaction_mass_balance](https://memote.readthedocs.io/en/latest/autoapi/test_consistency/index.html?highlight=charge#test_consistency.test_reaction_mass_balance).

**Reply by haowang-bioinfo** [reply]
might be a good idea to reuse the memote tests, and probably begin with mass and charge balance checks?

**Reply by haowang-bioinfo** [reply]
> it would also be nice if we can check wether the added rxns are mass balanced and charge balanced. I can work on that.

@feiranl agree and feel free to go ahead with your checks, regardless of memote tests

**Reply by feiranl** [reply]
I suggest performing a balance check for the model after each update. Regarding the output and failure criteria, my suggestion is to generate a default file documenting the currently unbalanced reactions and store it in a test folder. For the update, the document is read and a new test is run, if new unbalanced reactions occur, it would indicate a check failure.

**Reply by haowang-bioinfo** [reply]
some thoughts IMHO:
- memote tests and in-house tests can go ahead together, then maybe decide which to use later
- output of such checks should be boolean by default, in the case of unbalanced/duplicated rxns that probably can be directly reported as error message (instead of a standalone file)

**Comment by haowang-bioinfo** [comment]
close for now, feel free to reopen

**Reply by mihai-sysbio** [reply]
While PRs #563 and #564 resolve the main discussion items, I feel some more actionable conclusions were needed before closing the discussion, for example:
- is there an intention to include some memote tests?
- how is the gene essentiality going to be reported?
- is there a desire to run the gene essentiality automatically?

**Reply by haowang-bioinfo** [reply]
IMHO:

> is there an intention to include some memote tests?

open to this

> how is the gene essentiality going to be reported?

now is reported when pushing new release to `main` branch

> is there a desire to run the gene essentiality automatically?

no, not yet - because computational load is too heavy

**Reply by mihai-sysbio** [reply]
> > how is the gene essentiality going to be reported?
> 
> now is reported when pushing new release to `main` branch

Are you thinking of posting this directly in the PR discussion? My concern is that it is hard to find.

> > is there a desire to run the gene essentiality automatically?
> 
> no, not yet - because computational load is too heavy

Even if it's heavy, if it's only run every few months then it's still good. There is an attempt to achieve this in PR #675.

**Comment by haowang-bioinfo** [comment]
>>> how is the gene essentiality going to be reported?
>> 
>> now is reported when pushing new release to `main` branch
>
> Are you thinking of posting this directly in the PR discussion? My concern is that it is hard to find.

any suggestions?

>>> is there a desire to run the gene essentiality automatically?
>> 
>> no, not yet - because computational load is too heavy
>
> Even if it's heavy, if it's only run every few months then it's still good. There is an attempt to achieve this in PR #675.

then try and see

**Reply by mihai-sysbio** [reply]
> any suggestions?

How about generating a chart (image) to embed directly in the Readme?
